### ISSUE TITLE
"current state" changes the message payload and topic

### ISSUE THREAD

**Issue by Spartan-II-117** [AYapejian/node-red-contrib-home-assistant#15]
the current state node does not pass the initial topic and payload.

**Comment by flamingm0e** [comment]
I just came here to post this bug.  I am seeing the entire msg get modified, and it is screwing with chatbot

**Comment by Spartan-II-117** [comment]
maybe what we need is a separate "check state" node that will pass the original payload

**Comment by jonlidgard** [comment]
I'm having the same issue. I'm setting a msg.source variable & it isn't getting through the 'current state' node.

**Comment by nordlead2005** [comment]
the current state node uses the input purely as a trigger and configuration. The original message is dropped and a new message is passed on.

I guess if I wanted a new node, or new functionality it would essentially combine the "check state" node with a "switch" node as a simple halt on match is insufficient for some use cases.

There are workarounds that you can make work typically using either join nodes and javascript or a handful of nodes and stoptimer, but they both have the potential to drop messages that should have gone through if the input event rate is too fast due to relying on timers or message counts.

**Comment by AYapejian** [comment]
Yes, this was by design because I had use cases of exactly one user (me) when I started haha.  That said I hadn't been using node-red very long at all and I can certainly see how this functionality is counter to what one might in more complicated flows.

I'm uncertain on what @Spartan-II-117 is asking for specifically, personally I like keeping the raw state value on the payload as it's easy to access and most frequently needed ( for me anyways ).  I could change this in any number of ways like not using payload using a msg.state or something like that, using payload like it is now but rewritting the original payload to `payloadOriginal`, burying all homeassistant nodes output to a consistent `payload.ha` namespace property.  Each has it's benefits/drawbacks and I'm sure there are other ways of dealing with this. I can't see this being a unique node-red concern, anyone already know how other libs handle the problem or have their own thoughts?

As it stands now I'm going to mark this as an enhancement request and we can use the thread for further discussion if you're still interested. If we come to some kind of consensus then I can make those changes.  In addition any changes should include standardizing all the nodes to a consistent API as I'd like to not have to make a breaking change like this again if it can be avoided ( however it would be nice to have all nodes conversing with the same schema before I go ahead and implement more features I have in my head ;) )

Lastly for the current state / switch idea, I like it and was thinking something on the same lines awhile back as it'd be helpful, if you'd like @nordlead2005 create a new request for it and we can discuss there, I think it should be kept seperate from this discussion though.

_NOTE: Properties and values that are `snake_case'd` are original props straight from HA ( `entity_id`, `old_state`, `new_state` )_

---
#### Current state of nodes inputs/outputs

* Events: All
  * Input: None, incoming request from home assistant
  * Output example:
```javascript
{
    event_type: "state_changed"
    topic:      "state_changed"
    payload:    {
        entity_id: "sensor.mqtt_dev_sensor"
        old_state: {...}
        new_state: {...}
    }
    _msgid: "39381a52.0679d6"
}
```

* Events: State Changed:
  * Input: None, incoming request from home assistant
  * Output Example
```javascript
{
    topic:   "sensor.mqtt_dev_sensor"
    payload: "121"
    data:    {
        entity_id: "sensor.mqtt_dev_sensor"
        old_state: {...}
        new_state: {...}
    }
    _msgid: "b3d34a36.f79e38"
}
```

* Current State:
  * Input: `payload.entity_id `
  * Example
```javascript
{
    topic:   "binary_sensor.mqtt_dev_binary_sensor"
    payload: "off"
    data:    {
        entity_id:    "binary_sensor.mqtt_dev_binary_sensor"
        state:        "off"
        attributes:   {...}
        last_changed: "2018-01-28T01:58:56.644938+00:00"
        last_updated: "2018-01-28T01:58:56.644938+00:00"
    }
    _msgid: "c0e83dfe.d4417"
}
```
    
* Get Template:
  * Input: `msg.template `
  * Output Example:
```javascript
{
    topic:    ""
    payload:  "sensor.mqtt_dev_sensor=199,"
    template: "{% for state in states.sensor %}{{ state.entity_id }}={{ state.state }},{% endfor %}"
    _msgid:   "ee212aef.528928"
}
```

* Get History:
  * Input: `msg.start_date `
  * Output Example:
```javascript
{
    topic:     ""
    payload:   '<history data>'
    startdate: '<date used in history fetch call>'
    _msgid:    "ee212aef.528928"
}
```

**Comment by darchambault** [comment]
I've run into issues with this precise behavior on more complicated flows. It seems that the "current state" node creates an entirely new `msg` object, which is a bit of an anti-pattern for nodes as it (among other things) loses all properties including the `msg._id` (which gets reset to a new value when going through the "current state" node).

My suggestion is that the "current state" node should set properties to the `msg` object as it sees fit, but by doing all its operations on the original `msg` object rather than creating an entirely new object. That way properties would be preserved, or if there is naming conflict can be renamed ahead of time for later use.
